README.md
# animation_loading_button

![Message](https://img.shields.io/badge/just%20the%20message-8A2BE2)
[![Awesome Flutter](https://img.shields.io/badge/Awesome-Flutter-blue.svg)](https://github.com/Solido/awesome-flutter)
![GitHub downloads](https://img.shields.io/github/downloads/letienkhang/progessloadingbutton/total.svg)
![GitHub issues](https://img.shields.io/github/issues/letienkhang/progessloadingbutton.svg)
![GitHub watchers](https://img.shields.io/github/watchers/letienkhang/progessloadingbutton.svg)

**animation_loading_button** is a Flutter package with a simple implementation of an animated loading button, complete with success animations.

![Gif demo](https://github.com/letienkhang/progessloadingbutton/assets/44312440/8d5cec20-0331-4770-a242-e139d3444692)

## Getting Started
Follow these steps to use this package

### Add dependency

```yaml
dependencies:
  animation_loading_button: ^0.0.3
```

### Add import package

```dart
import 'package:animation_loading_button/animation_loading_button.dart';
```
### Easy to use
Simple example of use loading button<br>
Put this code in your project at an screen and learn how it works 😊
```dart
LoadingButton(
              type: LoadingButtonType.color,
              // Content inside the button when the button state is idle.
              idleStateWidget: const Text(
                'Color button',
                style: TextStyle(
                  color: Colors.black,
                ),
              ),
              useEqualLoadingStateWidgetDimension: true,
              // Whether or not to animate the width of the button. Default is `true`.
              // If this is set to `false`, you might want to set the `useEqualLoadingStateWidgetDimension` parameter to `true`.
              useAnimation: true,
              loadingType: LoadingType.circleSpinIndicator,
              // If you want a fullwidth size, set this to double.infinity
              width: 150.0,
              height: 40.0,
              buttonColor: Colors.blueAccent,
              loadingColor: Colors.white,
              onPressed: (){},
            ),
```

## Source
Source code and example of this library can be found in git:
```
$ git clone https://github.com/letienkhang/progessloadingbutton.git
```



CHANGELOG.md
## 0.0.3

Tiny Changes
* Update name project

## 0.0.2

Tiny Changes
* Update readme & demo

## 0.0.1

First usable release.


test/loadingbutton_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:loadingbutton/loadingbutton.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:loadingbutton/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/loadingbutton.dart
library loadingbutton;


import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/svg.dart';

class LoadingButton extends StatefulWidget {
  const LoadingButton({
    Key? key,
    required this.idleStateWidget,
    this.type = LoadingButtonType.color,
    this.loadingType = LoadingType.cupertinoActivityIndicator,
    this.useAnimation = true,
    this.useEqualLoadingStateWidgetDimension = true,
    this.width = double.infinity,
    this.height = 40.0,
    this.contentGap = 6.0,
    this.borderRadius = 4.0,
    this.elevation = 2.0,
    this.buttonColor = Colors.blueAccent,
    this.loadingColor = Colors.amber,
    this.onPressed,
  }) : super(key: key);

  final LoadingButtonType type;
  final LoadingType loadingType;
  final Widget idleStateWidget;
  final bool useAnimation;
  final bool useEqualLoadingStateWidgetDimension;
  final double width;
  final double height;
  final double contentGap;
  final double borderRadius;
  final double elevation;
  final Color buttonColor;
  final Color loadingColor;
  final Function? onPressed;

  @override
  State<LoadingButton> createState() => _LoadingButtonState();
}

class _LoadingButtonState extends State<LoadingButton>
    with TickerProviderStateMixin {
  final GlobalKey _globalKey = GlobalKey();

  Animation? _anim;
  AnimationController? _animController;
  final Duration _duration = const Duration(
    milliseconds: 250,
  );
  LoadingButtonState _state = LoadingButtonState.idle;
  late double _width;
  late double _height;
  late double _borderRadius;

  @override
  dispose() {
    if (_animController != null) {
      _animController!.dispose();
    }

    super.dispose();
  }

  @override
  void deactivate() {
    _reset();
    super.deactivate();
  }

  @override
  void initState() {
    _reset();
    super.initState();
  }

  void _reset() {
    _state = LoadingButtonState.idle;
    _width = widget.width;
    _height = widget.height;
    _borderRadius = widget.borderRadius;
  }

  @override
  Widget build(BuildContext context) {
    return PhysicalModel(
      color: Colors.transparent,
      borderRadius: BorderRadius.circular(_borderRadius),
      child: SizedBox(
        key: _globalKey,
        height: _height,
        width: _width,
        child: _buildChild(context),
      ),
    );
  }

  _buildChild(BuildContext context) {
    var padding = EdgeInsets.all(
      widget.contentGap,
    );
    var buttonColor = widget.buttonColor;
    var shape = RoundedRectangleBorder(
      borderRadius: BorderRadius.circular(_borderRadius),
    );

    final ButtonStyle elevatedButtonStyle = ElevatedButton.styleFrom(
      padding: padding,
      backgroundColor: buttonColor,
      elevation: widget.elevation,
      shape: shape,
    );

    final ButtonStyle outlinedButtonStyle = OutlinedButton.styleFrom(
      padding: padding,
      shape: shape,
      side: BorderSide(
        color: buttonColor,
      ),
    );

    final ButtonStyle textButtonStyle = TextButton.styleFrom(
      padding: padding,
    );

    switch (widget.type) {
      case LoadingButtonType.color:
        return ElevatedButton(
          style: elevatedButtonStyle,
          onPressed: _onButtonPressed(),
          child: _buildChildren(context),
        );
      case LoadingButtonType.transparent:
        return TextButton(
          style: outlinedButtonStyle,
          onPressed: _onButtonPressed(),
          child: _buildChildren(context),
        );
      case LoadingButtonType.text:
        return TextButton(
          style: textButtonStyle,
          onPressed: _onButtonPressed(),
          child: _buildChildren(context),
        );
    }
  }

  Widget _buildChildren(BuildContext context) {
    double contentGap =
        widget.type == LoadingButtonType.text ? 0.0 : widget.contentGap;
    Widget contentWidget;

    switch (_state) {
      case LoadingButtonState.idle:
        contentWidget = widget.idleStateWidget;

        break;
      case LoadingButtonState.loading:
        contentWidget = _switchLoadingType(widget.loadingType);

        if (widget.useEqualLoadingStateWidgetDimension) {
          contentWidget = SizedBox.square(
              dimension: widget.height - (contentGap * 2),
              child: _switchLoadingType(widget.loadingType));
        }

        break;

      case LoadingButtonState.done:
        contentWidget =
            widget.loadingType == LoadingType.circularProgressIndicator
                ? _circularProgressIndicator()
                : _cupertinoActivityIndicator();
        break;
    }

    return contentWidget;
  }

  void _reverse() {
    _animController!.reverse();
  }

  _circularProgressIndicator() {
    return CircularProgressIndicator(
      valueColor: AlwaysStoppedAnimation<Color>(
        widget.loadingColor,
      ),
      strokeWidth: 3.0,
    );
  }

  _cupertinoActivityIndicator() {
    return CupertinoActivityIndicator(
      color: widget.loadingColor,
    );
  }

  _sleekCircularSlider() {
    return SvgPicture.asset(
      'assets/images/loading.svg',
      semanticsLabel: 'A shark?!',
      allowDrawingOutsideViewBox: true,
      colorFilter: ColorFilter.mode(widget.loadingColor, BlendMode.srcIn),
    );
  }

  _circularRingIndicator() {
    return SvgPicture.asset(
      'assets/images/loading2.svg',
      semanticsLabel: 'A beef',
      allowDrawingOutsideViewBox: true,
      colorFilter: ColorFilter.mode(widget.loadingColor, BlendMode.srcIn),
    );
  }

  _switchLoadingType(LoadingType type) {
    Widget widget = _circularProgressIndicator();
    switch (type) {
      case LoadingType.circularProgressIndicator:
        widget = _circularProgressIndicator();
        break;
      case LoadingType.cupertinoActivityIndicator:
        widget = _circularProgressIndicator();
        break;
      case LoadingType.circleSpinIndicator:
        widget = _sleekCircularSlider();
        break;
      case LoadingType.circleRingIndicator:
        widget = _circularRingIndicator();
        break;
    }
    return widget;
  }

  void _forward(AnimationStatusListener stateListener) {
    double initialWidth = _globalKey.currentContext!.size!.width;
    double initialBorderRadius = widget.borderRadius;
    double targetWidth = _height;
    double targetBorderRadius = _height / 2;

    _animController = AnimationController(duration: _duration, vsync: this);
    _anim = Tween(begin: 0.0, end: 1.0).animate(_animController!)
      ..addListener(() {
        setState(() {
          _width = initialWidth - ((initialWidth - targetWidth) * _anim!.value);
          _borderRadius = initialBorderRadius -
              ((initialBorderRadius - targetBorderRadius) * _anim!.value);
        });
      })
      ..addStatusListener(stateListener);

    _animController!.forward();
  }

  void _toDefault() {
    if (mounted) {
      setState(() {
        _state = LoadingButtonState.idle;
      });
    } else {
      _state = LoadingButtonState.idle;
    }
  }

  void _toProcessing() {
    setState(() {
      _state = LoadingButtonState.loading;
    });
  }

  VoidCallback _onButtonPressed() {
    if (widget.onPressed == null) {
      return () {};
    }

    return _manageLoadingState;
  }

  Future _manageLoadingState() async {
    if (_state != LoadingButtonState.idle) {
      return;
    }
    dynamic onIdle;

    if (widget.useAnimation) {
      _toProcessing();
      _forward((status) {
        if (status == AnimationStatus.dismissed) {
          _toDefault();
          if (onIdle != null &&
              (onIdle is VoidCallback || onIdle is FormFieldValidator)) {
            onIdle();
          }
        }
      });
      onIdle = await widget.onPressed!();
      _reverse();
    } else {
      _toProcessing();
      onIdle = await widget.onPressed!();
      _toDefault();
      if (onIdle != null &&
          (onIdle is VoidCallback || onIdle is FormFieldValidator)) {
        onIdle();
      }
    }
  }
}

enum LoadingButtonState {
  idle,
  loading,
  done,
}

enum LoadingButtonType {
  color,
  transparent,
  text,
}

enum LoadingType {
  circularProgressIndicator,
  cupertinoActivityIndicator,
  circleSpinIndicator,
  circleRingIndicator
}



example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "loadingbutton");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "loadingbutton");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"loadingbutton", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/animation_loading_button.dart
library animation_loading_button;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/svg.dart';

class LoadingButton extends StatefulWidget {
  const LoadingButton({
    Key? key,
    required this.idleStateWidget,
    this.type = LoadingButtonType.color,
    this.loadingType = LoadingType.cupertinoActivityIndicator,
    this.useAnimation = true,
    this.useEqualLoadingStateWidgetDimension = true,
    this.width = double.infinity,
    this.height = 40.0,
    this.contentGap = 6.0,
    this.borderRadius = 4.0,
    this.elevation = 2.0,
    this.buttonColor = Colors.blueAccent,
    this.loadingColor = Colors.amber,
    this.onPressed,
  }) : super(key: key);

  final LoadingButtonType type;
  final LoadingType loadingType;
  final Widget idleStateWidget;
  final bool useAnimation;
  final bool useEqualLoadingStateWidgetDimension;
  final double width;
  final double height;
  final double contentGap;
  final double borderRadius;
  final double elevation;
  final Color buttonColor;
  final Color loadingColor;
  final Function? onPressed;

  @override
  State<LoadingButton> createState() => _LoadingButtonState();
}

class _LoadingButtonState extends State<LoadingButton>
    with TickerProviderStateMixin {
  final GlobalKey _globalKey = GlobalKey();

  Animation? _anim;
  AnimationController? _animController;
  final Duration _duration = const Duration(
    milliseconds: 250,
  );
  LoadingButtonState _state = LoadingButtonState.idle;
  late double _width;
  late double _height;
  late double _borderRadius;

  @override
  dispose() {
    if (_animController != null) {
      _animController!.dispose();
    }

    super.dispose();
  }

  @override
  void deactivate() {
    _reset();
    super.deactivate();
  }

  @override
  void initState() {
    _reset();
    super.initState();
  }

  void _reset() {
    _state = LoadingButtonState.idle;
    _width = widget.width;
    _height = widget.height;
    _borderRadius = widget.borderRadius;
  }

  @override
  Widget build(BuildContext context) {
    return PhysicalModel(
      color: Colors.transparent,
      borderRadius: BorderRadius.circular(_borderRadius),
      child: SizedBox(
        key: _globalKey,
        height: _height,
        width: _width,
        child: _buildChild(context),
      ),
    );
  }

  _buildChild(BuildContext context) {
    var padding = EdgeInsets.all(
      widget.contentGap,
    );
    var buttonColor = widget.buttonColor;
    var shape = RoundedRectangleBorder(
      borderRadius: BorderRadius.circular(_borderRadius),
    );

    final ButtonStyle elevatedButtonStyle = ElevatedButton.styleFrom(
      padding: padding,
      backgroundColor: buttonColor,
      elevation: widget.elevation,
      shape: shape,
    );

    final ButtonStyle outlinedButtonStyle = OutlinedButton.styleFrom(
      padding: padding,
      shape: shape,
      side: BorderSide(
        color: buttonColor,
      ),
    );

    final ButtonStyle textButtonStyle = TextButton.styleFrom(
      padding: padding,
    );

    switch (widget.type) {
      case LoadingButtonType.color:
        return ElevatedButton(
          style: elevatedButtonStyle,
          onPressed: _onButtonPressed(),
          child: _buildChildren(context),
        );
      case LoadingButtonType.transparent:
        return TextButton(
          style: outlinedButtonStyle,
          onPressed: _onButtonPressed(),
          child: _buildChildren(context),
        );
      case LoadingButtonType.text:
        return TextButton(
          style: textButtonStyle,
          onPressed: _onButtonPressed(),
          child: _buildChildren(context),
        );
    }
  }

  Widget _buildChildren(BuildContext context) {
    double contentGap =
        widget.type == LoadingButtonType.text ? 0.0 : widget.contentGap;
    Widget contentWidget;

    switch (_state) {
      case LoadingButtonState.idle:
        contentWidget = widget.idleStateWidget;

        break;
      case LoadingButtonState.loading:
        contentWidget = _switchLoadingType(widget.loadingType);

        if (widget.useEqualLoadingStateWidgetDimension) {
          contentWidget = SizedBox.square(
              dimension: widget.height - (contentGap * 2),
              child: _switchLoadingType(widget.loadingType));
        }

        break;

      case LoadingButtonState.done:
        contentWidget =
            widget.loadingType == LoadingType.circularProgressIndicator
                ? _circularProgressIndicator()
                : _cupertinoActivityIndicator();
        break;
    }

    return contentWidget;
  }

  void _reverse() {
    _animController!.reverse();
  }

  _circularProgressIndicator() {
    return CircularProgressIndicator(
      valueColor: AlwaysStoppedAnimation<Color>(
        widget.loadingColor,
      ),
      strokeWidth: 3.0,
    );
  }

  _cupertinoActivityIndicator() {
    return CupertinoActivityIndicator(
      color: widget.loadingColor,
    );
  }

  _sleekCircularSlider() {
    return SvgPicture.asset(
      'assets/images/loading.svg',
      semanticsLabel: 'A shark?!',
      allowDrawingOutsideViewBox: true,
      colorFilter: ColorFilter.mode(widget.loadingColor, BlendMode.srcIn),
    );
  }

  _circularRingIndicator() {
    return SvgPicture.asset(
      'assets/images/loading2.svg',
      semanticsLabel: 'A beef',
      allowDrawingOutsideViewBox: true,
      colorFilter: ColorFilter.mode(widget.loadingColor, BlendMode.srcIn),
    );
  }

  _switchLoadingType(LoadingType type) {
    Widget widget = _circularProgressIndicator();
    switch (type) {
      case LoadingType.circularProgressIndicator:
        widget = _circularProgressIndicator();
        break;
      case LoadingType.cupertinoActivityIndicator:
        widget = _circularProgressIndicator();
        break;
      case LoadingType.circleSpinIndicator:
        widget = _sleekCircularSlider();
        break;
      case LoadingType.circleRingIndicator:
        widget = _circularRingIndicator();
        break;
    }
    return widget;
  }

  void _forward(AnimationStatusListener stateListener) {
    double initialWidth = _globalKey.currentContext!.size!.width;
    double initialBorderRadius = widget.borderRadius;
    double targetWidth = _height;
    double targetBorderRadius = _height / 2;

    _animController = AnimationController(duration: _duration, vsync: this);
    _anim = Tween(begin: 0.0, end: 1.0).animate(_animController!)
      ..addListener(() {
        setState(() {
          _width = initialWidth - ((initialWidth - targetWidth) * _anim!.value);
          _borderRadius = initialBorderRadius -
              ((initialBorderRadius - targetBorderRadius) * _anim!.value);
        });
      })
      ..addStatusListener(stateListener);

    _animController!.forward();
  }

  void _toDefault() {
    if (mounted) {
      setState(() {
        _state = LoadingButtonState.idle;
      });
    } else {
      _state = LoadingButtonState.idle;
    }
  }

  void _toProcessing() {
    setState(() {
      _state = LoadingButtonState.loading;
    });
  }

  VoidCallback _onButtonPressed() {
    if (widget.onPressed == null) {
      return () {};
    }

    return _manageLoadingState;
  }

  Future _manageLoadingState() async {
    if (_state != LoadingButtonState.idle) {
      return;
    }
    dynamic onIdle;

    if (widget.useAnimation) {
      _toProcessing();
      _forward((status) {
        if (status == AnimationStatus.dismissed) {
          _toDefault();
          if (onIdle != null &&
              (onIdle is VoidCallback || onIdle is FormFieldValidator)) {
            onIdle();
          }
        }
      });
      onIdle = await widget.onPressed!();
      _reverse();
    } else {
      _toProcessing();
      onIdle = await widget.onPressed!();
      _toDefault();
      if (onIdle != null &&
          (onIdle is VoidCallback || onIdle is FormFieldValidator)) {
        onIdle();
      }
    }
  }
}

enum LoadingButtonState {
  idle,
  loading,
  done,
}

enum LoadingButtonType {
  color,
  transparent,
  text,
}

enum LoadingType {
  circularProgressIndicator,
  cupertinoActivityIndicator,
  circleSpinIndicator,
  circleRingIndicator
}


example/lib/main.dart
import 'package:animation_loading_button/animation_loading_button.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  onButtonPressed() async {
    await Future.delayed(const Duration(milliseconds: 3000), () => 42);

    // After [onPressed], it will trigger animation running backwards, from end to beginning
    return () {
      // Optional returns is returning a VoidCallback that will be called
      // after the animation is stopped at the beginning.
      // A best practice would be to do time-consuming task in [onPressed],
      // and do page navigation in the returned VoidCallback.
      // So that user won't missed out the reverse animation.
    };
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            LoadingButton(
              type: LoadingButtonType.color,
              // Content inside the button when the button state is idle.
              idleStateWidget: const Text(
                'Color button',
                style: TextStyle(
                  color: Colors.black,
                ),
              ),
              useEqualLoadingStateWidgetDimension: true,
              // Whether or not to animate the width of the button. Default is `true`.
              // If this is set to `false`, you might want to set the `useEqualLoadingStateWidgetDimension` parameter to `true`.
              useAnimation: true,
              loadingType: LoadingType.circleSpinIndicator,
              // If you want a fullwidth size, set this to double.infinity
              width: 150.0,
              height: 40.0,
              buttonColor: Colors.blueAccent,
              loadingColor: Colors.white,
              onPressed: onButtonPressed,
            ),
            const SizedBox(
              height: 10,
            ),
            LoadingButton(
              type: LoadingButtonType.transparent,
              // Content inside the button when the button state is idle.
              idleStateWidget: const Text(
                'Color button',
                style: TextStyle(
                  color: Colors.black,
                ),
              ),
              useEqualLoadingStateWidgetDimension: true,
              // Whether or not to animate the width of the button. Default is `true`.
              // If this is set to `false`, you might want to set the `useEqualLoadingStateWidgetDimension` parameter to `true`.
              useAnimation: true,
              loadingType: LoadingType.cupertinoActivityIndicator,
              // If you want a fullwidth size, set this to double.infinity
              width: 150.0,
              height: 40.0,
              buttonColor: Colors.blueAccent,
              loadingColor: Colors.black,
              onPressed: onButtonPressed,
            ),
          ],
        ),
      ),
    );
  }
}
